cmd/soroban-cli/src/commands/network/start.rs

```diff
@@ -0,0 +1,151 @@
+use std::process::{Command, Stdio};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {}
+
+const FROM_PORT: i32 = 8000;
+const TO_PORT: i32 = 8000;
+const CONTAINER_NAME: &str = "stellar";
+const DOCKER_IMAGE: &str = "stellar/quickstart";
+
+/// This command allows for starting a stellar quickstart container. To run it, you can use the following command:
+/// `soroban network start <NETWORK> [OPTIONS] -- [DOCKER_RUN_ARGS]`
+///
+/// OPTIONS: refer to the options that are available to the quickstart image:
+/// --enable-soroban-rpc - is enabled by default
+/// --protocol-version (only for local network)
+/// --limits (only for local network)
+
+/// `DOCKER_RUN_ARGS`: These are arguments to be passed to the `docker run` command itself, and should be passed in after the slop `--`. Some common options are:
+/// -p <`FROM_PORT`>:<`TO_PORT`> - this maps the port from the container to the host machine. By default, the port is 8000.
+/// -d - this runs the container in detached mode, so that it runs in the background
+
+// By default, without any optional arguments, the following docker command will run:
+// docker run --rm -p 8000:8000 --name stellar stellar/quickstart:testing --testnet --enable-soroban-rpc
+
+#[derive(Debug, clap::Parser, Clone)]
+pub struct Cmd {
+    /// Network to start, e.g. local, testnet, futurenet, pubnet
+    pub network: String,
+
+    /// optional argument to override the default docker image tag for the given network
+    #[arg(short = 't', long)]
+    pub image_tag_override: Option<String>,
+
+    /// optional argument to turn off soroban rpc
+    #[arg(short = 'r', long)]
+    pub disable_soroban_rpc: bool,
+
+    // option argument to specify the protocol version for the local network only
+    #[arg(short = 'p', long)]
+    pub protocol_version: Option<String>,
+
+    // option argument to specify the limits for the local network only
+    #[arg(short = 'l', long)]
+    pub limit: Option<String>,
+
+    // optional arguments to pass to the docker run command
+    #[arg(last = true, id = "DOCKER_RUN_ARGS")]
+    pub slop: Vec<String>,
+}
+
+impl Cmd {
+    pub fn run(&self) -> Result<(), Error> {
+        println!("Starting {} network", &self.network);
+
+        let docker_command = build_docker_command(self);
+
+        run_docker_command(&docker_command);
+        Ok(())
+    }
+}
+
+fn run_docker_command(docker_command: &str) {
+    println!("Running docker command: `{docker_command}`");
+    let mut cmd = Command::new("sh")
+        .args(["-c", &docker_command])
```

Comment:
Running docker commands will not only require a docker daemon, but a docker compatible CLI. While most container runtimes do provide docker compatible CLIs.

The daemon has a standardized API we should be able to call instead of shelling out.

I have less experience with this in the land of Rust, but in Go CLIs that use docker it's common to just import the docker libs and use them to talk to the daemon directly.

I would think there's a way we could do this in Rust too.

Possibly one of these projects would be similar. I haven't used any though:
- https://github.com/softprops/shiplift
- https://github.com/fussybeaver/bollard

Reply:
Great call! I was able to get this working with bollard in commit [51fbdce](https://github.com/stellar/soroban-tools/pull/1107/commits/51fbdcee45a9d43af11d7bcd8c6fa90884aae9c3) for start, and [198c070](https://github.com/stellar/soroban-tools/pull/1107/commits/198c07040921307723972d78176fa3aac1bc93e7) for stop.

I use docker desktop, and learned that it puts the docker socket file in a different location than the expected default. I was able to find a couple of ways around this that I'll make sure to document somewhere before merging this in. But the TLDR; is that there is now a `--docker-socket-path` flag the user can pass in the socket location on their machine.